### PR TITLE
Reduce CacheEntry size

### PR DESCRIPTION
**Pull request by adamsitnik**
Based on a customer feedback received in https://github.com/dotnet/runtime/issues/45592#issuecomment-741994471 I decided to try to shrink `CacheEntry` a little bit more. I was able to reduce the managed memory needed for allocating a `CacheEntry` from 280 to 224 bytes (on x64).

Main changes:

- remove two enum fields (`2x sizeof(int)`), store the information in an union struct called `CacheEntryState`. It now consists of two bytes for the mentioned enums and one byte for all the boolean flags. There is one more that is still not used, so we can add more stuff in the future ;). This looks like a magic and @roji is responsible for teaching me this kind of crazy stuff ;)
- move 3 lists related to tokens to a separate type called `CacheEntryTokens`. This allows us for saving space if they are not used (the most common scenario)
- remove the `_lock` field, use the field that stores `CacheEntryTokens` for locking instead (it guards the access to the tokens and the lock is needed only to work with the tokens)

The CPU time has not changed, but allocations did:

|                           Method |                Toolchain | Allocated |
|--------------------------------- |------------------------- |----------:|
|                                  |                          |           |
|                      SetOverride |  \after_size\CoreRun.exe |     224 B |
|                      SetOverride | \before_size\CoreRun.exe |     280 B |
|                                  |                          |           |
|                      CreateEntry |  \after_size\CoreRun.exe |     224 B |
|                      CreateEntry | \before_size\CoreRun.exe |     280 B |
|                                  |                          |           |
|       AddThenRemove_NoExpiration |  \after_size\CoreRun.exe |  42,370 B |
|       AddThenRemove_NoExpiration | \before_size\CoreRun.exe |  47,058 B |
|                                  |                          |           |
| AddThenRemove_AbsoluteExpiration |  \after_size\CoreRun.exe |  41,265 B |
| AddThenRemove_AbsoluteExpiration | \before_size\CoreRun.exe |  47,490 B |
|                                  |                          |           |
| AddThenRemove_RelativeExpiration |  \after_size\CoreRun.exe |  54,482 B |
| AddThenRemove_RelativeExpiration | \before_size\CoreRun.exe |  61,090 B |
|                                  |                          |           |
|  AddThenRemove_SlidingExpiration |  \after_size\CoreRun.exe |  41,313 B |
|  AddThenRemove_SlidingExpiration | \before_size\CoreRun.exe |  46,002 B |
|                                  |                          |           |
|   AddThenRemove_ExpirationTokens |  \after_size\CoreRun.exe |  53,706 B |
|   AddThenRemove_ExpirationTokens | \before_size\CoreRun.exe |  54,634 B |

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryTokens.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Extensions.Caching.Memory
+{
+    internal partial class CacheEntry
+    {
+        // this type exists just to reduce average CacheEntry size
+        // which typicall is not using expiration tokens or callbacks
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryTokens.cs:13]:
```suggestion
        // which typically is not using expiration tokens or callbacks
```

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryState.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryState.cs:1]: > What do you think about aligning the name of these files with the full class name?

I think it's a good idea! However, I am not sure which name would be the best: `CacheEntry.CacheEntryState.cs` or `CacheEntry.State.cs`?

> Either that or don't nest the Type and just use regular internal Types.

I would prefer to keep them `private` (and nested)

[the remark this answers is not in the document]
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryState.cs:1]: > I am not sure which name would be the best: CacheEntry.CacheEntryState.cs or CacheEntry.State.cs?

My stronger opinion is that file names and class names match. So whatever you decide for the class name, use that for the file name.

[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryState.cs]: done

**Comment by hopperpl** [comment]
Question: why is the atomic interlocked-exchange needed? 

If you update the whole 4-byte field, it needs to be atomic, yes. But if you just update the individual field, shouldn't the JIT guarantee that only the partial bytes are changed in an atomic way? 

```
            internal EvictionReason EvictionReasonSimple
            {
                get => (EvictionReason)_evictionReason;
                set
                {
                        _evictionReason = (byte)value;
                }
            }
```

results in ...
```
Microsoft.Extensions.Caching.Memory.CacheEntry+CacheEntryState.set_EvictionReasonSimple(Microsoft.Extensions.Caching.Memory.EvictionReason)
    L0000: mov [rcx+1], dl
    L0003: ret

```

which is just a single byte update... and does not change the surrounding bytes. It's an atomic operation on the CPU. Fields can be changed independently and don't interfere with another. 

I would get rid of _state and update the fields normally. The JIT should account for the correct atomic field update.

Unless I'm overlooking something here.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            private byte _reserved; // for future use
+
+            internal CacheEntryState(CacheItemPriority priority) : this() => _priority = (byte)priority;
+
+            internal bool IsDisposed { get => ((Flags)_flags).HasFlag(Flags.IsDisposed); set => SetFlag(Flags.IsDisposed, value); }
+
+            internal bool IsExpired
+            {
+                [MethodImpl(MethodImplOptions.AggressiveInlining)] // added based on profiling
+                get => ((Flags)_flags).HasFlag(Flags.IsExpired);
```
Comment (jkotas) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:37]:
```suggestion
                get => ((Flags)_flags & Flags.IsExpired) != 0;
```
This will inline nicely on its own, and avoids a ton of busy work in the JIT to get rid of the boxing and method call that `Enum.HasFlag` compiles into.

Comment (jkotas) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:37]: Same for all other HasFlag uses in this file.

**Comment by adamsitnik** [comment]
> Question: why is the atomic interlocked-exchange needed?

@hopperpl I've analysed the code again (and the way state can be updated) and you were right, it was not needed. I've added a simple test to ensure that.

**Comment by hopperpl** [comment]
> 
> 
> > Question: why is the atomic interlocked-exchange needed?
> 
> @hopperpl I've analysed the code again (and the way state can be updated) and you were right, it was not needed. I've added a simple test to ensure that.

If you need to avoid a race condition when different flags are set/unset at the same time, use Interlocked.Or, Interlocked.And (on the 32-bit _state value). This atomic operation won't interfere with the other fields.

```
  private void SetFlag(Flags option, bool value)
  {
    // _flags must be at offset 0, otherwise shift "options" into position
    // also needs a BigEndian platform check

    if (value)
      Interlocked.Or(ref _state, (byte) option);
    else 
      Interlocked.And(ref _state, (byte) ~option);
  }

```

**Comment by adamsitnik** [comment]
> If you need to avoid a race condition when different flags are set/unset at the same time, use Interlocked.Or, Interlocked.And

@hopperpl thanks! I've analysed the code and there is no need for it (and this particular lib targets .NET Standard 2.0 and these methods were introduced in .NET 5)

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                set => _evictionReason = (byte)value;
+            }
+
+            internal CacheItemPriority Priority
+            {
+                get => (CacheItemPriority)_priority;
+                set => _priority = (byte)value;
+            }
+
+            private void SetFlag(Flags option, bool value) => _flags = (byte)(value ? (_flags | (byte)option) : (_flags & ~(byte)option));
```
Comment (gfoidl) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: It's atomic (https://github.com/dotnet/runtime/pull/45962#issuecomment-743388974), but what about the visibility of the change? I think it should be volatile read and writes. Also for the properties here.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs [hunk 5]
```diff
@@ -11,28 +11,26 @@
...
     {
         private static readonly Action<object> ExpirationCallback = ExpirationTokensExpired;
 
-        private readonly object _lock = new object();
         private readonly MemoryCache _cache;
 
-        private IList<IDisposable> _expirationTokenRegistrations;
-        private IList<PostEvictionCallbackRegistration> _postEvictionCallbacks;
-        private IList<IChangeToken> _expirationTokens;
+        private CacheEntryTokens _tokens; // might be null if user is not using the tokens or callbacks
```
Comment (gfoidl) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs:20]:
```suggestion
        private CacheEntryTokens? _tokens; // might be null if user is not using the tokens or callbacks
```
Is nullability enabled in this project?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs (hunk 4 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: the `IsValueSet` and `IsDisposed` flags are set (and read) by a single thread that adds the entry to the cache, the `IsExpired` can, in theory, be set by two threads (one that is performing `cache.TryGetValue` and another that performs the background check for expired items) but even if the change is not visible to the other thread, it will just compare the dates and set `IsExpired` to true again (the `CheckExpired` logic)

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs (hunk 5 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs:20]: it is not

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        // this type exists just to reduce average CacheEntry size
+        // which typically is not using expiration tokens or callbacks
+        private sealed class CacheEntryTokens
+        {
+            internal List<IChangeToken> _expirationTokens;
+            internal List<IDisposable> _expirationTokenRegistrations;
+            internal List<PostEvictionCallbackRegistration> _postEvictionCallbacks; // this is not really related to tokens, but was moved here to shrink typicall CacheEntry size
+
+            internal List<IChangeToken> ExpirationTokens => _expirationTokens ??= new List<IChangeToken>();
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:23]: good point, I've made the fields private

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                        }
+                    }
+                }
+            }
+
+            internal bool CheckForExpiredTokens(CacheEntry cacheEntry)
+            {
+                if (_expirationTokens != null)
+                {
+                    for (int i = 0; i < _expirationTokens.Count; i++)
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:53]: @gfoidl I am not the author of the original code and I am not 100% sure, but to me, it looks that you are right:

- `AttachTokens` and `DetachTokens` are taking the lock to protect the access to `_expirationTokenRegistrations` (we are safe here)
- `CopyTokens` takes two locks to protect `_expirationTokens` of the `parent` cache entry and it looks that `CopyTokens` can modify `_expirationTokens` while it might be used by unprotected `CheckForExpiredTokens` at the same time

I've filled https://github.com/dotnet/runtime/issues/46032 to make sure it gets addressed

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs (hunk 4 above)
Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: > `IsExpired` can, in theory, be set by two threads (one that is performing `cache.TryGetValue` and another that performs the background check for expired items) but even if the change is not visible to the other thread

The "double write" is not the problem. I see the problem when it comes to reading the value, as the fresh value may not be visible then.

Let's assume Thread A on CPU 1 [sets expired](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L163-L169) and the [bits of _flags are set](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs#L35).
Thread B on CPU 2 [checks for expiration](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L158-L159) which [reads the corresponding bit of _flags](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs#L34).

What guarantees that Thread B sees the freshest value of `_flags`? 
It could happen that the change in _flags gets trapped in CPU 1's cache, so that CPU 2 / Thread B see a stale value.

I believe especially on ARM with the weak memory model this could be a concern.
Yeah, `volatile` has a perf hit on ARM, but it's needed for correctness?
Instead of the `volatile`-field I'd use `Volatile.Read` and `Volatile.Write` (to make it more obvious what happens).
Maybe there's a better way this can be handled -- at the moment I have no (other) idea (except Interlocked, but that's even heavier, also for x86/x64).

[ScanForExpiredItems](https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L311) check is concurrent, and this won't change with https://github.com/dotnet/runtime/pull/45842 [this call](https://github.com/dotnet/runtime/blob/14b3513cbe00e8306d305f7bcef6a5ae8d58acea/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L288).

Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: > What guarantees that Thread B sees the freshest value of _flags?

There is no such guarantee.

In this case, thread `B` might get `false` from `_state.IsExpired` so it's going to perform the full check:

https://github.com/dotnet/runtime/blob/c3e75eac361ee87e4bc2265d7fb303d5933f0978/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L160

and find out that it's expired anyway. Moreover, as soon as cache entry is set to expired by any thread it also gets removed from the cache a moment later. So we are talking about an extra very rare check

```ini
// thread A                 Thread B
------------------------|---------------------------
entry.SetExpired()      |   entry.IsExpired 
entries.Remove(entry)   |   entry.CheckExpired
                        |   entries.Remove(entry)
```

Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: So in other words: `_state.IsExpired` is quick-check-hint to avoid the full check. In the case of a race, we fall back to the full check.
Is this correct? If so, then there's no problem.

Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: @gfoidl exactly! thank you for a great discussion

Comment (gfoidl) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: Thank you too! Always nice to discuss about memory models.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Extensions.Caching.Memory
+{
+    internal partial class CacheEntry
+    {
+        // this type exists just to reduce CacheEntry size by replacing many enum & boolean fields with one of a size of Int32
+        [StructLayout(LayoutKind.Explicit)]
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:12]: Why does this need to have explicit layout?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                    {
+                        for (int i = 0; i < _expirationTokens.Count; i++)
+                        {
+                            IChangeToken expirationToken = _expirationTokens[i];
+                            if (expirationToken.ActiveChangeCallbacks)
+                            {
+                                if (_expirationTokenRegistrations == null)
+                                {
+                                    _expirationTokenRegistrations = new List<IDisposable>(1);
+                                }
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:40]:
```suggestion
                                _expirationTokenRegistrations ??= new List<IDisposable>(1);
```

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            [FieldOffset(0)]
+            private byte _flags;
+            [FieldOffset(1)]
+            private byte _evictionReason;
+            [FieldOffset(2)]
+            private byte _priority;
+            [FieldOffset(3)]
+            private byte _reserved; // for future use
+
+            internal CacheEntryState(CacheItemPriority priority) : this() => _priority = (byte)priority;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:24]: `: this()` doesn't seem necessary.

```suggestion
            internal CacheEntryState(CacheItemPriority priority) => _priority = (byte)priority;
```

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        private struct CacheEntryState
+        {
+            [FieldOffset(0)]
+            private byte _flags;
+            [FieldOffset(1)]
+            private byte _evictionReason;
+            [FieldOffset(2)]
+            private byte _priority;
+            [FieldOffset(3)]
+            private byte _reserved; // for future use
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:22]: Why is this necessary? Wouldn't just having 3 byte fields do the same thing?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+                                _expirationTokenRegistrations.Add(registration);
+                            }
+                        }
+                    }
+                }
+            }
+
+            internal bool CheckForExpiredTokens(CacheEntry cacheEntry)
+            {
+                if (_expirationTokens != null)
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:51]: Why doesn't this need a lock like the other two methods?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+
+            internal bool CanCopyTokens() => _expirationTokens != null;
+
+            internal void CopyTokens(CacheEntry parentEntry)
+            {
+                if (_expirationTokens != null)
+                {
+                    lock (this)
+                    {
+                        lock (parentEntry.GetOrCreateTokens())
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:74]: Where else is this locked on?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+                        {
+                            foreach (IChangeToken expirationToken in _expirationTokens)
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:76]: Why in some cases are we using a for loop with direct indexing to enumerate the contents and in others we're using foreach?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+            internal void DetachTokens()
+            {
+                // _expirationTokenRegistrations is not checked for null, because AttachTokens might initialize it under lock
+                // instead we are checking for _expirationTokens, because if they are not null, then _expirationTokenRegistrations might also be not null
+                if (_expirationTokens != null)
+                {
+                    lock (this)
+                    {
+                        IList<IDisposable> registrations = _expirationTokenRegistrations;
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:93]: Why IList rather than List?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+                if (_postEvictionCallbacks != null)
+                {
+                    Task.Factory.StartNew(state => InvokeCallbacks((CacheEntry)state), cacheEntry,
+                        CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
+                }
+            }
+
+            private static void InvokeCallbacks(CacheEntry entry)
+            {
+                IList<PostEvictionCallbackRegistration> callbackRegistrations = Interlocked.Exchange(ref entry._tokens._postEvictionCallbacks, null);
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:118]: Why IList rather than List?

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs (hunk 4 above)
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:56]: Why don't we need to use `Interlocked.CompareExchange` here? Won't we potentially lose changes if 1 thread is setting `IsExpired` while another thread is setting `IsDisposed`?

Note: we were doing that in the previous code:
https://github.com/dotnet/runtime/blob/5078dd61a81c88dbdd131a77b8148b4dee3803fa/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L352-L361

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs [hunk 17]
```diff
@@ -262,60 +212,11 @@ private static void ExpirationTokensExpired(object obj)
...
-                    entry._cache._logger.LogError(e, "EvictionCallback invoked failed");
-                }
-            }
-        }
+        internal void InvokeEvictionCallbacks() => _tokens?.InvokeEvictionCallbacks(this);
 
         // this simple check very often allows us to avoid expensive call to PropagateOptions(CacheEntryHelper.Current)
-        internal bool CanPropagateOptions() => _expirationTokens != null || AbsoluteExpiration.HasValue;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // added based on profiling
+        internal bool CanPropagateOptions() => (_tokens != null && _tokens.CanCopyTokens()) || AbsoluteExpiration.HasValue;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs:219]: (nit) Can we rename `_tokens.CanCopyTokens()` => `_tokens.CanPropagateTokens()`? It feels odd calling these operations 2 different words (`Propagate` and `Copy`) in the same line.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs (hunk 8 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:12]: The initial implementation had one in32 field that was overlapping four byte fields, but I've removed it in https://github.com/dotnet/runtime/pull/45962/commits/46ae2eadcaeed0e35166fe44302c1264413784bd and forgot to remove the attribute. I am going to do it right away

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs (hunk 10 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs]: obraz

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs (hunk 4 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:49]: Other code reviewers asked why it was needed, I've analyzed the code again and it turned out to be not needed (https://github.com/dotnet/runtime/pull/45962#issuecomment-743333631)

> 1 thread is setting IsExpired while another thread is setting IsDisposed

The unusual thing about this particular `Dispose` is that it's being called to... add the entry to the cache. So nobody can set it as expired at the same time because it has not been added to the cache yet

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs (hunk 12 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:48]: https://github.com/dotnet/runtime/pull/45962#discussion_r542216496

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs (hunk 14 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:73]: I don't know as I am not the author of the original code. To minimize the diff and make the review easier I've decided to change as few things as possible

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs (hunk 13 above)
Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:71]: https://github.com/dotnet/runtime/pull/45962#discussion_r542216496

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:71]: I'm not sure how that answers my question.  My primary concern would be if it'd ever be possible for something to take the second lock and then try to take the first, creating a priority inversion and deadlock potential.  I'm simply wondering if that's possible.  Does the child/parent nature of this guarantee locks are always acquired in the same order?

Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:71]: > Does the child/parent nature of this guarantee locks are always acquired in the same order?

I think it does in the case when the cache item is being added to the cache (by calling the Dispose method):

https://github.com/dotnet/runtime/blob/6e111ea8695635624147227cf685fd6769b88ea4/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L155-L157

but I am not 100% sure if it would not be possible to create a code that does this and abuses the second possible call to this method from the `Cache.TryGetValue` method (which is very unintuitive for me):

https://github.com/dotnet/runtime/blob/6e111ea8695635624147227cf685fd6769b88ea4/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L237-L239

FWIW I have not changed the locking order, it's the same as it used to be so far:

https://github.com/dotnet/runtime/blob/6e111ea8695635624147227cf685fd6769b88ea4/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.cs#L331-L340

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:49]: The object is handed out to consumers, right? Could someone set the value and call Dispose on it from 2 different threads?

Comment (adamsitnik) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:49]: > The object is handed out to consumers, right? Could someone set the value and call Dispose on it from 2 different threads?

In theory yes, but in general we don't have any thread safe `Dispose` methods in BCL? Personally I don't see the value in the additional complexity.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryState.cs:49]: I just called it out since there is so much other thread safety concerns here (since this is a cache).

If you are satisfied with it, that works for me.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+{
+    internal partial class CacheEntry
+    {
+        // this type exists just to reduce average CacheEntry size
+        // which typically is not using expiration tokens or callbacks
+        private sealed class CacheEntryTokens
+        {
+            private List<IChangeToken> _expirationTokens;
+            private List<IDisposable> _expirationTokenRegistrations;
+            private List<PostEvictionCallbackRegistration> _postEvictionCallbacks; // this is not really related to tokens, but was moved here to shrink typicall CacheEntry size
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntry.CacheEntryTokens.cs:21]:
```suggestion
            private List<PostEvictionCallbackRegistration> _postEvictionCallbacks; // this is not really related to tokens, but was moved here to shrink typical CacheEntry size
```

**Review by eerhardt — APPROVED** [review]
LGTM. Nice improvement here!
